Document/0x06d-Testing-Data-Storage.md

```diff
@@ -761,16 +761,53 @@ This will cause the background image to be set to the "overlayImage.png" instead
+
+```swift
+var str1 = "Goodby"              // "Goodby", base adress:             0x0001039e8dd0
+str1.append(" ")                 // "Goodby ", base adress:            0x608000064ae0
+str1.append("cruel world!")      // "Goodby cruel world", base adress: 0x6080000338a0
+str1.removeAll()                 // "", base adress                    0x00010bd66180
+```
+
+Notice how the base address of the underlying value changes with each modification. The problem here is that in order to securely erase the sensitive information from memory we don't want to simply change the value of the variable, but the actual content of the memory that the current value uses. _Swift_ does not offer such API.
+
+_Swift_ collections (`Array`, `Set` and `Dictionary`), on another hand, might be acceptable as long as they collect primitive data types such as `char` or `int` and are defined as mutable (variables instead of constants), in which case they are more or less equivalent to a primitive array (like `char []`). Still, these collections provide safe memory management, which can result in unidentified copies of the sensitive data in memory, in case the collection needs to copy the underlying buffer to a different location in order to extend it.
+
+Usage of _Objective-C_ mutable data types, such as `NSMutableString` might also be acceptable, but suffer from the same "safe memory" issue as _Swift_ collections. Additionally, attention should be payed when using _Objective-C_ collections as they hold data by reference and only _Objective-C_ data types are allowed. Therefore, we are not looking for a mutable collection, but a collection that references mutable objects.
+
+As we've seen so far usage of _Swift_ or _Objective-C_ data types require deep understanding of the language implementation itself. Further, we have witnessed some core re-factoring between mayor _Swift_ versions, resulting in incompatible behavior of many data types. In order to avoid these issues we recommend usage of primitive data types whenever certain data needs to be securely erased from memory.
+
+Unfortunately, not many libraries and frameworks are designed to allow overwriting of sensitive data. Even Apple does not consider this issue in the official API of the iOS SDK. For example, most of the APIs for data transformation (passers, serializes, etc.) operate on non-primitive data types. Similarly, regardless of weather you flag certain `UITextField` as _Secure Text Entry_ or not, it always return data as `String` or `NSString`.
 
+To summarize, when performing static analysis for sensitive data exposed in memory, you should:
+- Try to identify application components and make a map of where certain data is used.
```

Comment:
> where certain data is used

I am not sure you can have "certain data", if you consider data an uncountable noun.